src/sentry/release_health/metrics.py

```diff
@@ -606,3 +605,336 @@ def extract_row_info(row: Mapping[str, Union[OrganizationId, str]]) -> ReleaseNa
+                        Op.EQ,
+                        session_init_tag_value,
+                    ),
+                ],
+                granularity=Granularity(stats_rollup),  # type: ignore
+                groupby=aggregates,
+            ),
+            referrer="release_health.metrics.get_health_stats_for_overview",
+        )["data"]:
+            time_bucket = int(
+                (parse_snuba_datetime(row["bucketed_time"]) - stats_start).total_seconds()
+                / stats_rollup
+            )
+            key = row["project_id"], reverse_tag_value(org_id, row[release_column_name])
+            timeseries = rv[key]
+            if time_bucket < len(timeseries):
+                timeseries[time_bucket][1] = row["value"]
+
+        return rv
+
+    def get_release_health_data_overview(
+        self,
+        project_releases: Sequence[ProjectRelease],
+        environments: Optional[Sequence[EnvironmentName]] = None,
+        summary_stats_period: Optional[StatsPeriod] = None,
+        health_stats_period: Optional[StatsPeriod] = None,
+        stat: OverviewStat = None,
+    ):
+        if stat is None:
+            stat = "sessions"
+        assert stat in ("sessions", "users")
+        now = datetime.now(pytz.utc)
+        _, summary_start, _ = get_rollup_starts_and_buckets(summary_stats_period or "24h")
+
+        org_id = self._get_org_id([x for x, _ in project_releases])
+
+        where: List[Union[BooleanCondition, Condition]] = [
+            Condition(Column("org_id"), Op.EQ, org_id),
+            filter_projects_by_project_release(org_id, project_releases),
+            Condition(Column("timestamp"), Op.GTE, summary_start),
+            Condition(Column("timestamp"), Op.LT, now),
+        ]
+
+        if environments is not None:
+            where.append(
+                Condition(
+                    Column(tag_key(org_id, "environment")),
+                    Op.IN,
+                    get_tag_values_list(org_id, environments),
+                )
+            )
+
+        if health_stats_period:
+            health_stats_data = self._get_health_stats_for_overview(
+                where, org_id, health_stats_period, stat, now
+            )
+        else:
+            health_stats_data = {}
+
+        rv_durations = self._get_session_duration_data_for_overview(where, org_id)
+        rv_errored_sessions = self._get_errored_sessions_for_overview(where, org_id)
+        rv_sessions = self._get_abnormal_and_crashed_sessions_for_overview(where, org_id)
+        rv_users = self._get_users_and_crashed_users_for_overview(where, org_id)
+
+        # XXX: In order to be able to dual-read and compare results from both
+        # old and new backend, this should really go back through the
+        # release_health service instead of directly calling `self`. For now
+        # that makes the entire backend too hard to test though.
+        release_adoption = self.get_release_adoption(project_releases, environments)
```

Comment:
Wouldn't that mean that a dual-read call to `get_release_health_data_overview` would call `get_release_adoption` 4 times (assuming that the "sessions" impl also calls `release_health.get_release_adoption`). IMO only the top-level calls (by the application) should be subject to dual-read.

Reply:
we can still decide whether to actually dual-read later. We can use a thread-local to track if it's an internal call or not

Reply:
BTW: this line does **not** call the releasehealth service. It calls the specific backend class.

If we have a dual-reader, `self` would still be the metrics backend. I assume the dual-reader would be a new class that contains two backend instances, not a class that is a subclass of both backends.